x-pack/osquerybeat/ext/osquery-extension/pkg/hooks/hooks.go

```diff
@@ -0,0 +1,95 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package hooks
+
+import (
+	"sync"
+
+	"github.com/elastic/beats/v7/x-pack/osquerybeat/ext/osquery-extension/pkg/logger"
+)
+
+// HookFunc represents a function that executes a hook, any
+// function that implements this interface can be used as a hook
+type HookFunc func(socket *string, log *logger.Logger, hookData any) error
+
+// Hook is a struct that represents a hook function
+type Hook struct {
+	hookName     string
+	hookFunc     HookFunc
+	shutdownFunc HookFunc
+	hookData     any
+}
+
+// Name returns the name of the hook
+func (h *Hook) Name() string {
+	return h.hookName
+}
+
+// Execute executes the hook function
+func (h *Hook) Execute(socket *string, log *logger.Logger) error {
+	if h.hookFunc == nil {
+		return nil
+	}
+	log.Infof("executing hook, name: %s", h.Name())
+	return h.hookFunc(socket, log, h.hookData)
+}
+
+func (h *Hook) Shutdown(socket *string, log *logger.Logger) error {
+	if h.shutdownFunc == nil {
+		return nil
+	}
+	log.Infof("executing shutdown hook, name: %s", h.Name())
+	return h.shutdownFunc(socket, log, h.hookData)
+}
+
+// NewHook creates a new hook
+func NewHook(hookName string, hookFunc HookFunc, shutdownFunc HookFunc, hookData any) *Hook {
+	return &Hook{hookName: hookName, hookFunc: hookFunc, shutdownFunc: shutdownFunc, hookData: hookData}
+}
+
+// HookManager is a struct that contains all hooks of a given type
+type HookManager struct {
+	hooks []*Hook
```

Comment:
nit: probably not needed given our use of it, but a mutex to protect racey Register/Execute calls would be nice